File: Chapter_16_-_Convolution/python/mach_illusion.py
# -*- coding: utf-8 -*-
#mach_illusion.m

import numpy as np
from scipy import signal
import matplotlib.pyplot as plt

#reset the variables before executing

from mexican_hat import * #creates the Mexican hat matrix, MH, & plots
from ramp import * #creates image with ramp from dark to light, stimulus, & plots
A = signal.convolve2d(stimulus,MH,'valid')  #convolve image and Mexican hat

plt.figure()
plt.imshow(A, cmap='Blues_r') #visualize the "perceived" brightness
plt.title('Perceived',fontsize=22)

#create plot showing the profile of both the input and the perceived brightness
plt.figure()
plt.plot(stimulus[31,:],'k',linewidth=5, label='input brightness')
plt.axis([0,128,-20,120])
plt.plot(A[31,:],'b-.',linewidth=2, label='perceived brightness') 
plt.legend()

plt.show()

File: Matlab_for_Neuroscientists/Chapter_16_-_Convolution/mach_illusion.m
%mach_illusion.m
clear all; close all
mexican_hat %creates the Mexican hat matrix, MH, & plots
ramp %creates image with ramp from dark to light, In, & plots
A = conv2(In,MH,'valid'); %convolve image and Mexican hat
figure; imagesc(A); colormap(bone) %visualize the "perceived" brightness
%create plot showing the profile of both the input and the perceived brightness
figure; plot(In(32,:),'k','LineWidth',5);axis([0,128,-20,120])
hold on; plot(A(32,:),'b-.','LineWidth',2);
lh = legend('input brightness','perceived brightness',2);